### ISSUE TITLE
[PROJECT2] Shop find function 

### ISSUE THREAD

**Issue by toastanddrink** [nus-cs2030/2021-s1#389]
Hi, can I check how do I go about implementing the find function for Shop? Do I have to use function and if so, may I check the types I should put into my parameter? 

So sorry for this vague question!

**Comment by b-leong** [comment]
you would use a predicate in the parameter, can convert a copy of the current list of servers into streams and use stream functions to get the desired return value

**Comment by Jcheez** [comment]
I agree with @b-leong. Do note that the return type of find should be an optional < Server > because there might be instances where there is no suitable server found.

**Comment by changziming** [comment]
Also keep in mind that you need to find the first available server.

**Comment by toastanddrink** [comment]
thank you guys for the help! I realised mine didn't work cuz I didn't find only the first server (:

**Comment by dljw98** [comment]
Hi, I basically used filter for f.apply(x), and added all Servers to a List<Server>. Then see if it's empty, and if not, use if-else to determine what output to return.

**Comment by bryanwee023** [comment]
@dljw98 Hmm correct me if I'm wrong but isn't finding every Server that fits the requirement redundant since we only have to find the first Server. As the comments above mentioned, I think it's smarter to convert Shop's list of Servers to a stream and exploit Stream's find method.

**Comment by Clem-cpu** [comment]
@dljw98 a good way to shorten that code is to use List<Server> s.stream().filter(f).findFirst(). This allows u to take advantage of the stream() findFirst method which helps do the if-else checking for you!!

**Comment by descentbw** [comment]
As what the rest have mentioned, make use of the stream methods, since the question requires Shop to be developed in a  functional style.

**Comment by kheekheekhee** [comment]
As was said previously, I believe that you'll have to use Streams and you can look into the firstFirst method that comes in Streams that is already built-in for you
